### ISSUE TITLE
"wait-for-db" init container fails on unavailable custom plugins

### ISSUE THREAD

**Issue by Tieske** [Kong/charts#371]
We're using a custom image that has custom plugins installed and in use.

When doing an upgrade the init container that wait for the postgres database to be available fails with;
```
2021/06/11 09:20:17 [error] 5276#0: init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:582: error building initial plugins: helloplugin plugin is in use but not enabled                                  ││ stack traceback:                                                                                                                                                                                                 ││     [C]: in function 'error'                                                                                                                                                                                     │
│     /usr/local/share/lua/5.1/kong/init.lua:582: in function 'init'                                                                                                                                               ││     init_by_lua:3: in main chunk                                                                                                                                                                                 ││   Run with --v (verbose) or --vv (debug) for more details                                                                                                                                                        ││ waiting for db                                                                                                                                                                                                   │
│ 2021/06/11 09:20:19 [warn] 5287#0: *2 [lua] plugins.lua:231: load_plugin(): plugin 'route-by-header' has been deprecated, context: ngx.timer
```

Quick workaround is adding an environment variable `KONG_PLUGINS=bundled,helloplugin`, but that should not be necessary since the docker image has start up scripts that do this already.

The reason is in this line: https://github.com/Kong/charts/blob/bd8e8600f708a956e4041ec84d6a0814e1fbcfd2/charts/kong/templates/_helpers.tpl#L507 where it does the regular `kong start` instead of using the docker entrypoint scripts located at `/docker-entrypoint.sh` in the docker container.

For reference, the entrypoint scripts: https://github.com/Kong/docker-kong/blob/master/alpine/docker-entrypoint.sh
And if customized with custom plugins this: https://github.com/Kong/docker-kong/blob/master/customize/packer.lua#L376-L411

fyi @mflendrich @lokesh-chechani

**Comment by rainest** [comment]
How are you loading the custom plugin elsewhere in values.yaml? It sounds like this isn't using the [dedicated plugin loader](https://github.com/Kong/charts/blob/kong-2.1.0/charts/kong/values.yaml#L251-L270) (which updates `KONG_PLUGINS`) and instead have manual mounts.

`wait-for-db` intentionally doesn't use the entrypoint script because we want the container to exit, whereas the entrypoint script will run indefinitely. If you're not using the dedicated plugin loader, manually setting `KONG_PLUGINS` is probably the correct approach.

You may be able to simply [disable the wait-for-db initContainer](https://github.com/Kong/charts/blob/kong-2.1.0/charts/kong/values.yaml#L437), as it's not strictly required (it's more a UX sugar thing to avoid a temporary crashloop state). Not sure how that will interact with the entrypoint script after.

**Comment by Tieske** [comment]
@rainest we used the [customize dockerfile](https://github.com/Kong/docker-kong/tree/master/customize) to add a custom plugin to a new image.

That dockerfile will automatically update entrypoint script to enable the plugin. So the image can be used like any other Kong image, no special settings required to enable the plugin. But to do that it does require running the entrypoint script.

**Comment by Tieske** [comment]
So this line actually is the problem:
https://github.com/Kong/docker-kong/blob/master/alpine/docker-entrypoint.sh#L29

Because it hard codes that setting, it should only set it if not set already. Because it should only override the default.

When running Kong it should not be daemonized, but when testing connectivity, it should be daemonized, such that on successful start the container exits immediately.

**Comment by Tieske** [comment]
see https://github.com/Kong/docker-kong/pull/474

**Comment by ttony** [comment]
@Tieske, instead of building your own custom image based on Kong. Have you considered using init-container solution to setup your custom plugin during the init phase?... once init-phase ready or done, then use job to enabled the plugin. that way your components are independent. 

FYI, i am also using custom plugins and i am using init-containers for this purpose.

**Comment by Tieske** [comment]
the "wait-for-db" container uses a bad entrypoint, it should use the script, which it can't. So we need a generic solution to this.

**Comment by rainest** [comment]
It wouldn't make any difference in this case--normal usage of that strategy still requires setting `KONG_PLUGINS` to make Kong aware of the additional plugins. The custom image builder here includes something that sets a default value other than `bundled` at build time, but that only works if using the included entrypoint script.

@Tieske how is this custom image system handling plugin migrations if you don't manually set `KONG_PLUGINS`? I know the traditional entrypoint script intercepted only the special `kong docker-start` command and ran in the foreground with that, whereas any other command would run as usual. Plugins that include migrations still need to be set in `KONG_PLUGINS` for runs of `kong migrations up`, so absent some custom handling of migrations commands or a new Docker-specific migrations command handled by the entrypoint script I don't think we'd be able to recommend that for general use (it'd be limited to use with zero-migration plugins, and although those are most common, they're not the entire ecosystem). There [doesn't appear to be anything in the new script that adds new migration intercepts](https://github.com/Kong/docker-kong/blob/2.4.1/customize/packer.lua#L395-L398).

For the chart specifically, if you do want to use this for no-migration plugins, I think my recommendation for now would still be to just disable the database waits using the values.yaml toggle. wait-for-db needs some way to check migrations completeness to continue serving its purpose, so the alternative would be to add a migrations readiness checker (or generic daemonized start command) in the new entrypoint script. While I was in favor of just removing it entirely the last time we encountered (unrelated) issues with it, but [others were opposed](https://github.com/Kong/charts/pull/283), so we plan to keep it in place and on by default for the foreseeable future.

**Comment by Tieske** [comment]
> There [doesn't appear to be anything in the new script that adds new migration intercepts](https://github.com/Kong/docker-kong/blob/2.4.1/customize/packer.lua#L395-L398).

That referenced lines are only for the `--nginx-conf` flag, since the other Kong commands will throw an error if you include that.

[These lines](https://github.com/Kong/docker-kong/blob/2.4.1/customize/packer.lua#L381-L387) will set the `KONG_PLUGINS` variable for any command, and hence cover the plugin migrations, right? (not sure how the helm charts test for migrations to be ran or not)

**Comment by rainest** [comment]
Possibly by intent, but apparently that's not working in practice? Do you see the `KONG_PLUGINS` you expect if you do `docker run -it image /bin/sh` and check the environment? What this does should be similar.

AFAIK the `command` in the container spec does _not_ (and in fact, cannot) override ENTRYPOINT: it's overriding CMD, which is passed as arguments to ENTRYPOINT. ENTRYPOINT here just happens to be a script that (mostly) executes whatever you pass it after some initial setup: https://github.com/Kong/docker-kong/blob/2.4.1/alpine/docker-entrypoint.sh#L50

The new entrypoint script is essentially passthrough to that after yet more setup and intercepts, and while my bash is terrible, the export should be fine:

```
14:16:15-0700 yagody $ export FOO="foo"
14:16:24-0700 yagody $ echo $FOO
foo
14:16:29-0700 yagody $ exec `which bash`
[traines@yagody ~]$ echo $FOO
foo
```
The export in the new script should still happen, and should then pass on to the environment that executes our `kong start` script in wait-for-db.

There is some additional complexity here because of https://github.com/Kong/charts/blob/kong-2.1.0/charts/kong/templates/_helpers.tpl#L699

The `autoenv` there uses `env.plugins` from the chart values if present, and if not, falls back to the `kong.plugins` helper output: https://github.com/Kong/charts/blob/kong-2.1.0/charts/kong/templates/_helpers.tpl#L487-L496

_That_ output will just be `bundled` if you're not using the chart's plugin loader system, but that should be fine since the new entrypoint script does still append to the end of existing  `KONG_PLUGINS` (which is what this is--the container env is equivalent to `-e` in `docker run`).

**Comment by Tieske** [comment]
> Possibly by intent, but apparently that's not working in practice? Do you see the KONG_PLUGINS you expect if you do docker run -it image /bin/sh and check the environment? What this does should be similar.

Yes I see them:
```
Thijss-MBP:customize thijs$ docker build \
>    --build-arg KONG_BASE="kong:latest" \
>    --build-arg PLUGINS="kong-http-to-https,kong-upstream-jwt" \
>    --build-arg "KONG_LICENSE_DATA=$KONG_LICENSE_DATA" \
>    --tag "your_new_image" .
[+] Building 12.3s (13/13) FINISHED
 => [internal] load build definition from Dockerfile                                                                  0.0s
 => => transferring dockerfile: 37B                                                                                   0.0s
 => [internal] load .dockerignore                                                                                     0.0s
 => => transferring context: 2B                                                                                       0.0s
 => [internal] load metadata for docker.io/library/kong:latest                                                        1.9s
 => [auth] library/kong:pull token for registry-1.docker.io                                                           0.0s
 => CACHED [build 1/5] FROM docker.io/library/kong:latest@sha256:c9872460e6c9cb05d5ac1f125fff571c00541d4da7438f5eb82  0.0s
 => => resolve docker.io/library/kong:latest@sha256:c9872460e6c9cb05d5ac1f125fff571c00541d4da7438f5eb821e07a56efe752  0.0s
 => [internal] load build context                                                                                     0.0s
 => => transferring context: 11.07kB                                                                                  0.0s
 => [build 2/5] COPY empty_file /plugins/custom_nginx.conf                                                            0.0s
 => [build 3/5] COPY empty_file /rocks-server                                                                         0.0s
 => [build 4/5] COPY packer.lua /packer.lua                                                                           0.0s
 => [build 5/5] RUN /usr/local/openresty/luajit/bin/luajit /packer.lua -- "kong-http-to-https,kong-upstream-jwt"      9.4s
 => [stage-1 2/3] COPY --from=build /plugins /plugins                                                                 0.0s
 => [stage-1 3/3] RUN /plugins/install_plugins.sh                                                                     0.4s
 => exporting to image                                                                                                0.0s
 => => exporting layers                                                                                               0.0s
 => => writing image sha256:1103de27c8d5f192f5b49804a5e2282cd3ed973d6d301cfd5b31ede1133473bb                          0.0s
 => => naming to docker.io/library/your_new_image                                                                     0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
Thijss-MBP:customize thijs$ docker run -it --rm your_new_image /bin/sh
/ $ printenv | grep KONG_
KONG_VERSION=2.4.1
KONG_AMD64_SHA=582cbcf23eb4dcdf9873105ac7d8428a4022ec61bcc68642ad9dd8a5c03e2a57
KONG_ARM64_SHA=5ec35d1b19dd4e6592ad2c6586e68bfd1549c6a22840ce7d5654677b94e5028a
KONG_NGINX_DAEMON=off
KONG_PLUGINS=bundled,http-to-https,kong-upstream-jwt
/ $
```

**Comment by Tieske** [comment]
@lokesh-chechani can you try the fix in https://github.com/Kong/charts/pull/377 ? and let us know if that works?

**Comment by rainest** [comment]
Closed presuming fix in #377. @lokesh-chechani please re-open if that change does not actually address the failure with image-bundled custom plugins.
